lib/copy-sync/copy-sync.js

```diff
@@ -68,10 +68,17 @@ function mayCopyFile (srcStat, src, dest, opts) {
 function copyFile (srcStat, src, dest, opts) {
   if (typeof fs.copyFileSync === 'function') {
     fs.copyFileSync(src, dest)
-    fs.chmodSync(dest, srcStat.mode)
     if (opts.preserveTimestamps) {
-      return utimesSync(dest, srcStat.atime, srcStat.mtime)
+      // Make sure the file is writable first
+      if ((srcStat.mode & 0o200) === 0) {
+        fs.chmodSync(dest, srcStat.mode | 0o200)
```

Comment:
Actually, it is required. One of the tests I added reveals that it doesn't work without this if the source file is read-only (manually tested on Windows and Mac).

1) If `src` is readonly, then so is `dest` after `fs.copyFileSync(src, dest)` 
2) `utimesSync` then fails with a permission error because it's not allowed to open the `dest` file for write.

Reply:
Simple repro on Node 8.11.3, with latest head (9f1c02916bb51d43feb7063f00f138e232f4d518). Assumed to be run from the fse directory.
```
const fse = require('.')
fse.ensureFileSync('src')
fse.chmodSync('src', 0x111)
fse.copySync('src', 'destNoPreserve') // works
fse.copySync('src', 'destPreserve', { preserveTimestamps: true }) // fails with the following stack trace (Windows example):
Error: EPERM: operation not permitted, open 'C:\Users\maximeb\code\node-fs-extra\destPreserve'
    at Object.fs.openSync (fs.js:646:18)
    at utimesMillisSync (C:\Users\maximeb\code\node-fs-extra\lib\util\utimes.js:68:17)
    at copyFile (C:\Users\maximeb\code\node-fs-extra\lib\copy-sync\copy-sync.js:79:7)
    at onFile (C:\Users\maximeb\code\node-fs-extra\lib\copy-sync\copy-sync.js:55:25)
    at getStats (C:\Users\maximeb\code\node-fs-extra\lib\copy-sync\copy-sync.js:50:44)
    at startCopy (C:\Users\maximeb\code\node-fs-extra\lib\copy-sync\copy-sync.js:40:10)
    at handleFilterAndCopy (C:\Users\maximeb\code\node-fs-extra\lib\copy-sync\copy-sync.js:35:10)
    at Object.copySync (C:\Users\maximeb\code\node-fs-extra\lib\copy-sync\copy-sync.js:28:10)
```

With this proposed change, no exception. If I comment out line 74, then I get the same exception.

If you don't want to keep this as-is, I see three options:
- Make sure `copyFileSync` does not set the `dest` mode to be the same as `src`
- Change `utimesMillisSync` to be more robust if it fails with EPERM
- Change `utimesMillisSync` to avoid this problem by not using `fs.openSync(<path>, 'r+')` + `fs.futimesSync` but straight out `fs.utimesSync` which works even on readonly files (note: I'm not sure the `futimes` and `utimes` interfaces are fully compatible, this might be the reason for using `futimes` in `utimesMillisSync` in the first place...)
- Change _this code_ to implement the chmod in a try/catch around utimesSync... but then it's no simpler than an `if`, just more costly.

Reply:
so this is my suggestion:

- do not set the dest mode when we create the dest
- use native `fs.utimes*()` and `fs.futimes*()` (instead of our internal `utimes` implementation)
- if we do the above, then we don't need to check if the file is writable and we can simplify the `setDestTimestampsAndMode` function like this, which is so much easier to read and understand. All tests also pass with this implementation. I tried your example too and that one worked as well:
```js
function setDestTimestampsAndMode (srcStat, src, dest, opts) {
  if (opts.preserveTimestamps) {
    const updatedSrcStat = fs.statSync(src)
    fs.utimesSync(dest, updatedSrcStat.atime, updatedSrcStat.mtime)
  }
  fs.chmodSync(dest, srcStat.mode)
}
```
```js
function setDestTimestampsAndModeFd (srcStat, src, fdw, opts) {
  if (opts.preserveTimestamps) {
    const updatedSrcStat = fs.statSync(src)
    fs.futimesSync(fdw, updatedSrcStat.atime, updatedSrcStat.mtime)
  }
  fs.fchmodSync(fdw, srcStat.mode)
}
```